### PR TITLE
Value based coordinate slicing (C++ implementation)

### PR DESCRIPTION
**Pull request by OwenArnold**
C++ implementation for value based slicing. Part of #1133

Python bindings should be relatively trivial but would be done as a follow up. As part of the bindings we may want to bind this free function to `DataArray` via pybind11 or alongside a wrapper python function, which we would likely need anyway (modern way to produce docs).

### CODE REVIEW

File: dataset/include/scipp/dataset/slice.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+/// @file
+/// @author Owen Arnold
+#pragma once
+
+#include "scipp/dataset/dataset.h"
+
+namespace scipp::dataset {
+
+[[nodiscard]] SCIPP_DATASET_EXPORT DataArrayConstView
+slice(const DataArrayConstView &to_slice, const Dim dim,
```
Comment (SimonHeybrock) [inline dataset/include/scipp/dataset/slice.h:12]: Also for `DatasetConstView`?

File: dataset/slice.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                         const VariableConstView end) {
+  using namespace variable;
+  if (begin)
+    variable::detail::expect0D(begin.dims());
+  if (end)
+    variable::detail::expect0D(end.dims());
+  auto coords = to_slice.coords();
+  if (!coords.contains(dim))
+    throw except::DimensionNotFoundError(to_slice.dims(), dim);
+  auto [k, coord] = *coords.find(dim);
```
Comment (SimonHeybrock) [inline dataset/slice.cpp:21]:
```suggestion
  auto coord = to_slice.coords()[dim];
```
?

File: dataset/slice.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    throw except::SizeError(
+        "Multi-dimensional coordinates not supported in slice");
+  }
+  const bool ascending = is_sorted(coord, dim, variable::SortOrder::Ascending);
+  const bool descending =
+      is_sorted(coord, dim, variable::SortOrder::Descending);
+
+  if (!(ascending ^ descending))
+    throw std::runtime_error("Coordinate must be monotomically increasing or "
+                             "decreasing for value slicing");
```
Comment (SimonHeybrock) [inline dataset/slice.cpp:32]: For non-range slice of point coords I think the order does not and should not matter by spectrum number?
```python
spec = data['spectrum', sc.Variable(value=17)]
```
?

File: dataset/slice.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  const auto len_data = to_slice.dims()[dim];
+  const auto len = coord.dims().volume();
+  const auto bin_edges = len == len_data + 1;
+
+  // Point slice
+  if ((begin && end) && begin == end) {
```
Comment (SimonHeybrock) [inline dataset/slice.cpp:39]: Not sure I understand the second part of the check, `begin == end` should return and empty range (dimension extent `0`), just like `data['x', 4:4]`.

File: dataset/slice.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+              len - sum(greater_equal(coord, begin), dim).value<scipp::index>();
+        else
+          first =
+              len - sum(less_equal(coord, begin), dim).value<scipp::index>();
+      }
+      if (first < 0)
+        first = 0;
+    }
+    if (end) {
+      if (ascending) {
```
Comment (SimonHeybrock) [inline dataset/slice.cpp:90]: Massive duplication around here. There should not be so many cases? For example, finding the index of bins is the same, regardless of whether it for begin or end (just need to add 1 to the index of the last bin). Also finding indices for non-bin case is the same, just the intepretation of the indices changes?

File: dataset/test/slice_test.cpp [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// SPDX-License-Identifier: GPL-3.0-or-later
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Comment (SimonHeybrock) [inline dataset/test/slice_test.cpp:1]: Typo?

File: dataset/slice.cpp (hunk 4 above)
Comment (SimonHeybrock) [reply dataset/slice.cpp:39]: I see you are trying to support `begin` and `end` unspecified by passing an invalid view? I think that is fine, but then there is no way to distinguish from a point slice, so maybe we need a differently named method for that?

**Comment by SimonHeybrock** [comment]
Note that I adressed my comments myself since @OwenArnold is on extended vacation.

**Comment by SimonHeybrock** [comment]
Apart from earlier comments, I also improved the behavior for non-range slices of point data:
- Fail for duplicate coord values.
- No sorting required.

**Comment by SimonHeybrock** [comment]
@nvaytet Could you give this a review? I made substantial change, so a third pair of eyes would be good. Cheers!

**Review by nvaytet — APPROVED** [review]
I think what you did with the two function signatures is much cleaner and more intuitive :+1:

**Pull request merged**